parsely/src/test/java/com/parsely/parselyandroid/EngagementManagerTest.kt

```diff
@@ -0,0 +1,158 @@
+
+        sleep(DEFAULT_INTERVAL_MILLIS)
+        val thirdTimestamp = now - 3 * THREAD_SLEEPING_THRESHOLD
+
+        sleep(THREAD_SLEEPING_THRESHOLD)
+
+        val firstEvent = tracker.events[0]
+        assertThat(firstEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(firstTimestamp, within(20L)) }
+        )
+        val secondEvent = tracker.events[1]
+        assertThat(secondEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS * 2
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS * 2, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(secondTimestamp, within(20L)) }
+        )
+        val thirdEvent = tracker.events[2]
+        assertThat(thirdEvent).isCorrectEvent(
+            // Ideally: totalTime should be equal to DEFAULT_INTERVAL_MILLIS * 3
+            withTotalTime = { isCloseTo(DEFAULT_INTERVAL_MILLIS * 3, withinPercentage(10)) },
+            // Ideally: timestamp should be equal to `now` at the time of recording the event
+            withTimestamp = { isCloseTo(thirdTimestamp, within(20L)) }
+        )
+    }
+
+    private fun MapAssert<String, Any>.isCorrectEvent(
+        withTotalTime: AbstractLongAssert<*>.() -> AbstractLongAssert<*>,
+        withTimestamp: AbstractLongAssert<*>.() -> AbstractLongAssert<*>,
+    ): MapAssert<String, Any> {
+        return containsEntry("action", "heartbeat")
+            // Incremental will be always 0 because the interval is lower than 1s
+            .containsEntry("inc", 0L)
+            .hasEntrySatisfying("tt") { totalTime ->
+                totalTime as Long
+                assertThat(totalTime).withTotalTime()
+            }
+            .hasEntrySatisfying("data") { data ->
+                data as Map<String, Any>
+                assertThat(data).hasEntrySatisfying("ts") { timestamp ->
+                    timestamp as Long
+                    assertThat(timestamp).withTimestamp()
+                }.containsAllEntriesOf(testData.minus("ts"))
+            }
+    }
+
+    class FakeTracker : ParselyTracker(
+        "",
+        0,
+        ApplicationProvider.getApplicationContext()
+    ) {
+        val events = mutableListOf<Event>()
+
+        override fun enqueueEvent(event: Event) {
+            events += event
+        }
+    }
+
+    class FakeIntervalCalculator : UpdateEngagementIntervalCalculator() {
+        override fun updateLatestInterval(startTime: Calendar): Long {
+            return DEFAULT_INTERVAL_MILLIS
+        }
+    }
+
+    private val now: Long
```

Comment:
Minor (🔍): Since this is a `val` field and not a method or an inner class, for better visibility, consider moving that at the top, alongside the rest of the fields.

Reply:
If that's okay, I'd leave this as it is. My reasoning: on top, we have declarations for the test subject and its dependencies. At the bottom: fakes and helper methods for test execution. `now` being `val` is only a syntax sugar to not add `()` in some places.